### DISCUSSION TITLE
NAs For Brier Skill Score

### DISCUSSION THREAD

**Discussion by craigrj** [dtcenter/METplus#1455] [Statistical Computation] [answered]
The following set up data used for computing a Brier Skill Score (BSS) of NA.  The BSS should be 1 since the Brier Score is 0 since the model probabilities were 1 for the threshold of >.01 in precip and each observation site also hit that threshold with at least .01" of precip.  The uncertainty term (
[listing.pdf](https://github.com/dtcenter/METplus/files/8143082/listing.pdf)
BSref)is 0.  This presents a divide by 0 in the BSS equation that should be handled.  Was this a problem on MET 8.1?  Or am I missing something
[listing.pdf](https://github.com/dtcenter/METplus/files/8143088/listing.pdf)

.

**Comment by JohnHalleyGotway** [comment]
@craigrj, I sifted carefully through the first line of the PDF you sent. As far as I can tell, here's the relevant values from the first PSTD line in that file:
```
BRIER   BRIERCL BSS
0.14145 NA      NA
```

The BSS is computed as 1.0 - (BRIER / BRIERCL), after checking for bad data on this [line of code](https://github.com/dtcenter/MET/blob/1a1da435712481c0b3b31db5042f78909d0a0f8d/met/src/libcode/vx_statistics/met_stats.cc#L2594). Since the climatological brier score has bad data (BRIERCL = NA), so too does BSS.

Note that the BRIERCL value is computed using the climatological data provided in the config file with climo_mean and climo_stdev.

Does that answer your question?

**Comment by craigrj** [comment]
I thought BSS Simple just used the frequency of the event for climo value - obar(1-obar).   BSS uses BrierCL.  So if the event didn’t occur, the frequency is 0 and if Brier is 0 (so the event wasn’t forecast to occur and it didn’t occur, isn’t the BSS 1 for that case?

**Reply by JohnHalleyGotway** [reply]
Looking at the [PSTD output table](https://met.readthedocs.io/en/latest/Users_Guide/point-stat.html#id15), just documenting more details.

- BSS in column 40 is the Brier Skill Score computed relative to the external climatology. You can see that computation for version 8.1 on [this line](https://github.com/dtcenter/MET/blob/0f939dcfc03864c25c7450fb3674141096779659/met/src/libcode/vx_statistics/met_stats.cc#L2599).
- BSS_SMPL in column 41 is the Brier Skill Score computed relative to the *sample* climatology. And the sample climatology is just the current set of observation values. You can see that computation for version 8.1 on [this line](https://github.com/dtcenter/MET/blob/0f939dcfc03864c25c7450fb3674141096779659/met/src/libcode/vx_statistics/contable_nx2.cc#L615). 

Currently, if the both the forecast brier score and climo brier score are 0, then BSS = 1 - 0 / 0, which results in divide by 0, and you'd see BSS = NA in the output.

Is your question really, if both the forecast and climo brier scores are 0, then *SHOULD* BSS actually be 1.0? The code is checking for and avoiding divide-by-zero. If this is your question, I can ask a statistician here more about it.

**Comment by craigrj** [comment, marked as answer]
Yes.  I understand the need to avoid divide by zero.  I guess the question is if the model doesn’t forecast the event and the event doesn’t occur, should this be a BSS of 1?  In reality, I am usually looking for BSS’s over a period of time and the chance of a sample climatology of 0 goes down with more cases.

**Reply by ericgilleland** [reply]
@craigrj @JohnHalleyGotway The original BS included both binary events but is generally now understood to be summed only over events (and sometimes called the half BS).  I'm not sure if that comes into play here or not.  I'm assuming MET computes the half BS?  Either way, yes, given that a perfect score for BS is 0, if both BS and BSref are zero, then the BSS should be 1.  I'm not sure what people do in practice, but it does present a problem going from perfectly 0 to say 1e-8 for one of them.  BTW-- BSref can be the BS for any reference forecast; I'm not sure how that works in MET, however.

**Reply by JohnHalleyGotway** [reply]
@ericgilleland thanks for taking a look. Here's the block of code in MET that computes BSS:
```
   //
   // Compute the brier skill score
   //
   if(is_bad_data(brier.v) || is_bad_data(briercl.v) ||
      is_eq(briercl.v, 0.0)) {
      bss = bad_data_double;
   }
   else {
      bss = 1.0 - (brier.v / briercl.v);
   }
```

If the climatological brier score is 0 (or very, very close to 0), we report BSS as bad data since we'd be dividing by 0. It sounds like you're saying we should modify this algorithm.

Should we add a check for when the forecast brier score is perfect (i.e. forecast brier score = 0), then report BSS = 1.0 regardless of the climo brier score? I'm looking for very clear direction on if/how this algorithm should be modified.

That'd look something like this:
```
   //
   // Compute the brier skill score
   //
   if(is_eq(brier.v, 0.0) && !is_bad_data(briercl.v)) {
      bss = 1.0;
   }
   else if(is_bad_data(brier.v) || is_bad_data(briercl.v) ||
      is_eq(briercl.v, 0.0)) {
      bss = bad_data_double;
   }
   else {
      bss = 1.0 - (brier.v / briercl.v);
   }
```

**Reply by ericgilleland** [reply]
@JohnHalleyGotway @craigrj The more I think about it, I think the reason I haven't seen anything about BSref's being 0 is that there is an assumption that BSref will never be perfect (so never zero).  Perhaps instead of NA, some kind of flag could be returned or an error message?  I guess I'm suggesting that the reference forecast shouldn't be perfect because you're already comparing them against perfection.  I will send an email to vx-discuss to see what the other experts say, and let you know if there is a consensus.  Maybe this idea has already been thoroughly digested.

**Reply by ericgilleland** [reply]
@craigrj @JohnHalleyGotway I got lots of responses on vx-discuss and I would go with David Stephenson's reply, which is to code it as NaN.  In R it means 0/0 as opposed to NA, which means not available.  So, if you don't use NaN with MET, then maybe just keep it as NA.  He agreed with what you said, Craig, that the only way to get BSref = 0 all values need to either be 0 or 1 because o(1-o) = 0.

**Reply by JohnHalleyGotway** [reply]
@ericgilleland and @craigrj, I realize that we never marked this discussion as being answered. Based on the feedback that Eric gathered, MET should be writing out "NaN" in this case... which stands for "Not A Number". In all other cases of divide-by-zero, MET just reports "NA" instead, for "Not Available". So we use "NA" in 2 ways... for columns that really don't apply and also for math issues (divide-by-zero, square root of negative number, etc). I can see some benefit in differentiating between NA and NaN in the output, but that change would be much larger than just BSS. We'd need to introduce/handle that new flag value and update the code in many places to use it. We'd also need to update METviewer to correctly ignore NaN and run the risk of breaking custom user scripts.

Practically speaking, I think that change would have little impact on the verification process. Both NA and NaN values would be ignored when aggregating results in Stat-Analysis and would also be ignored when making plots in METviewer. Seeing no clear and obvious practical benefit, I'd prefer to just leave the output as-is... which is also the easiest thing to do.

If you do think MET should be enhanced to differentiate between NA and NaN in the output, please let me know. Otherwise, next Monday, I'll just mark this discussion as being answered and lock it to prevent future posts.

Thanks for the discussion.

**Comment by craigrj** [comment]
I was really asking your statistical experts if an event not forecast  and didn’t occur should that be scored as a 1.  If they tend to agree then the change you mentioned makes sense.  If they don’t, then don’t worry about it.  I appreciate the discussion on this issue.

**Reply by ericgilleland** [reply]
@craigrj I think that might get at the idea that the BS is only accumulated over events, but I'm not sure.  In principle, though, I would say yes.

**Comment by craigrj** [comment]
Thanks for the discussion.   I would be happy with whatever you guys decide is best.

**Comment by craigrj** [comment]
Okay, thanks.

**Comment by craigrj** [comment]
That okay with me.

 

Bob

 

From: johnhg ***@***.***> 
Sent: Tuesday, March 29, 2022 10:17 AM
To: dtcenter/METplus ***@***.***>
Cc: CRAIG, ROBERT J GS-12 USAF ACC 16 WS/WXD ***@***.***>; Mention ***@***.***>
Subject: [URL Verdict: Neutral][Non-DoD Source] Re: [dtcenter/METplus] NAs For Brier Skill Score (Discussion #1455)

 

@ericgilleland <https://github.com/ericgilleland>  and @craigrj <https://github.com/craigrj> , I realize that we never marked this discussion as being answered. Based on the feedback that Eric gathered, MET should be writing out "NaN" in this case... which stands for "Not A Number". In all other cases of divide-by-zero, MET just reports "NA" instead, for "Not Available". So we use "NA" in 2 ways... for columns that really don't apply and also for math issues (divide-by-zero, square root of negative number, etc). I can see some benefit in differentiating between NA and NaN in the output, but that change would be much larger than just BSS. We'd need to introduce/handle that new flag value and update the code in many places to use it. We'd also need to update METviewer to correctly ignore NaN and run the risk of breaking custom user scripts.

Practically speaking, I think that change would have little impact on the verification process. Both NA and NaN values would be ignored when aggregating results in Stat-Analysis and would also be ignored when making plots in METviewer. Seeing no clear and obvious practical benefit, I'd prefer to just leave the output as-is... which is also the easiest thing to do.

If you do think MET should be enhanced to differentiate between NA and NaN in the output, please let me know. Otherwise, next Monday, I'll just mark this discussion as being answered and lock it to prevent future posts.

Thanks for the discussion.

**Comment by ericgilleland** [comment]
Yeah, I don't see it as a big deal whether they are NA or NaN.
